core/src/test/java/org/elasticsearch/index/query/support/InnerHitBuilderTests.java

```diff
@@ -0,0 +1,255 @@
+            innerHits.setNestedPath(randomAsciiOfLengthBetween(1, 16));
+        } else {
+            innerHits.setParentChildType(randomAsciiOfLengthBetween(1, 16));
+        }
+
+        innerHits.setName(randomAsciiOfLengthBetween(1, 16));
+        innerHits.setFrom(randomIntBetween(0, 128));
+        innerHits.setSize(randomIntBetween(0, 128));
+        innerHits.setExplain(randomBoolean());
+        innerHits.setVersion(randomBoolean());
+        innerHits.setTrackScores(randomBoolean());
+        innerHits.setFieldNames(randomListStuff(16, () -> randomAsciiOfLengthBetween(1, 16)));
+        innerHits.setFieldDataFields(randomListStuff(16, () -> randomAsciiOfLengthBetween(1, 16)));
+        innerHits.setScriptFields(randomListStuff(16, InnerHitBuilderTests::randomScript));
+        FetchSourceContext randomFetchSourceContext;
+        if (randomBoolean()) {
+            randomFetchSourceContext = new FetchSourceContext(randomBoolean());
+        } else {
+            randomFetchSourceContext = new FetchSourceContext(
+                    generateRandomStringArray(12, 16, false),
+                    generateRandomStringArray(12, 16, false)
+            );
+        }
+        innerHits.setFetchSourceContext(randomFetchSourceContext);
+        if (randomBoolean()) {
+            innerHits.setSorts(randomListStuff(16,
+                    () -> SortBuilders.fieldSort(randomAsciiOfLengthBetween(5, 20)).order(randomFrom(SortOrder.values())))
+            );
+        }
+        innerHits.setHighlightBuilder(HighlightBuilderTests.randomHighlighterBuilder());
+        if (randomBoolean()) {
+            innerHits.setQuery(new MatchQueryBuilder(randomAsciiOfLengthBetween(1, 16), randomAsciiOfLengthBetween(1, 16)));
+        }
+        if (recursive && randomBoolean()) {
+            InnerHitsBuilder innerHitsBuilder = new InnerHitsBuilder();
+            int size = randomIntBetween(1, 16);
+            for (int i = 0; i < size; i++) {
+                innerHitsBuilder.addInnerHit(randomAsciiOfLengthBetween(1, 16), randomInnerHits(false));
+            }
+            innerHits.setInnerHitsBuilder(innerHitsBuilder);
+        }
+
+        return innerHits;
+    }
+
+    static InnerHitBuilder mutate(InnerHitBuilder innerHits) throws IOException {
+        InnerHitBuilder copy = serializedCopy(innerHits);
+        int suprise = randomIntBetween(0, 5);
+        switch (suprise) {
+            case 0:
+                copy.setFrom(Integer.MAX_VALUE);
+                break;
+            case 1:
+                copy.setSize(Integer.MAX_VALUE);
+                break;
+            case 2:
+                copy.setExplain(!copy.isExplain());
+                break;
+            case 3:
+                copy.setVersion(!copy.isVersion());
+                break;
+            case 4:
+                copy.setTrackScores(!copy.isTrackScores());
+                break;
+            case 5:
+                copy.setName("CHANGED");
+                break;
```

Comment:
Can this be expanded to a few more (if not all) potential properties of InnerHitBuilder? Maybe the more complex random generation of the parts can be factored out from randomInnerHits(). Theres also a helper in ESTestCase#randomValueOtherThan() that might help not accidentally picking the same value that was used in the original InnerHitBuilder.